packages/completer/src/tokens.ts

```diff
@@ -138,6 +138,131 @@ export interface ICompletionProvider<
   ): boolean;
 }
 
+/**
+ * Describes how an inline completion provider was triggered.
+ * @alpha
+ */
+export enum InlineCompletionTriggerKind {
+  /**
+   * Completion was triggered explicitly by a user gesture.
+   * Return multiple completion items to enable cycling through them.
+   */
+  Invoke = 0,
+  /**
+   * Completion was triggered automatically while editing.
+   * It is sufficient to return a single completion item in this case.
+   */
+  Automatic = 1
+}
+
+/**
+ * The context which will be passed to the inline completion provider.
+ * @alpha
+ */
+export interface IInlineCompletionContext {
+  /**
+   * The widget (notebook, console, code editor) which invoked
+   * the inline completer
+   */
+  widget: Widget;
+
+  /**
+   * The session extracted from widget for convenience.
+   */
+  session?: Session.ISessionConnection | null;
+
+  triggerKind: InlineCompletionTriggerKind;
+
+  // selectedCompletionInfo?: ISelectedCompletionInfo;
+}
+
+/**
+ * LSP 3.18-compliant inline completion API subset.
+ */
+interface IInlineCompletionItemLSP {
+  /**
+   * The text to replace the range with. Must be set.
+   * Is used both for the preview and the accept operation.
+   */
+  insertText: string;
+
+  /**
+   * A text that is used to decide if this inline completion should be
+   * shown. When `falsy` the insertText is used.
+   *
+   * An inline completion is shown if the text to replace is a prefix of the
+   * filter text.
+   */
+  filterText?: string;
+}
+
+/**
+ * An inline completion item represents a text snippet that is proposed inline
+ * to complete text that is being typed.
+ * @alpha
+ */
+export interface IInlineCompletionItem extends IInlineCompletionItemLSP {
+  token?: string;
+  /**
+   * Whether generation of `insertText` is still ongoing. If your provider supports streaming,
+   * you can set this to true, which will result in the provider's `stream()` method being called
+   * with `token` which has to be set for incomplete completions.
+   */
+  isIncomplete?: boolean;
+}
+
+export interface IInlineCompletionList<
+  T extends IInlineCompletionItem = IInlineCompletionItem
+> {
+  /**
+   * The inline completion items.
+   */
+  items: T[];
+}
+
+/**
+ * The interface to implement an inline completion provider.
+ */
+export interface IInlineCompletionProvider<
+  T extends IInlineCompletionItem = IInlineCompletionItem
+> {
+  fetch(
+    request: CompletionHandler.IRequest,
+    context: IInlineCompletionContext
+  ): Promise<IInlineCompletionList<T>>;
+
+  /**
+   * Name of the provider to be displayed in the UI.
+   */
+  name: string;
```

Comment:
I guess it is set at instantiation:

```suggestion
  readonly name: string;
```